### PR TITLE
Added reprojection and forced CRS handling to ContentFeatureSource

### PR DESCRIPTION
**Pull request by tbarsballe**
API Clarifications to Query javadoc.
Added forced CRS handling to ContentFeatureSource getBounds and getReader.
Added reprojection handling to getBounds.
Added test cases for forced CRS handling and reprojection to MemoryDataStore (the ContentFeatureSource test mocks are insufficient to test the changes to ContentFeatureSource). Also updated DataTestCase to use EPSG:4326 for the the river data set, to allow for testing tranformations and pre-assigned CRSs (Verified that this does not break any other test cases).

### CODE REVIEW

File: modules/library/data/src/main/java/org/geotools/data/store/ContentFeatureSource.java [hunk 1]
```diff
@@ -425,6 +427,35 @@ public final ReferencedEnvelope getBounds(Query query) throws IOException {
...
+                if(sourceCRS == null) {
+                    throw new IOException("Cannot reproject data, the source CRS is not available");
+                } else if(!sourceCRS.equals(targetCRS)) {
+                    try {
+                        bounds = bounds.transform(targetCRS, true);
+                    } catch (Exception e) {
+                        if(e instanceof IOException)
+                            throw (IOException) e;
+                        else
+                            throw (IOException) new IOException("Error occurred trying to reproject data").initCause(e);
```
Comment (jodygarnett) [inline modules/library/data/src/main/java/org/geotools/data/store/ContentFeatureSource.java]: Could probably use new IOException("Unable to reproject bounds",e)

File: modules/library/data/src/main/java/org/geotools/data/store/ContentFeatureSource.java [hunk 2]
```diff
@@ -425,6 +427,35 @@ public final ReferencedEnvelope getBounds(Query query) throws IOException {
         } else {
             bounds = getBoundsInternal(query);
         }
+        // reprojection
+        if ( !canReproject() ) {
+            CoordinateReferenceSystem sourceCRS = query.getCoordinateSystem();
+            CoordinateReferenceSystem targetCRS = query.getCoordinateSystemReproject();
+            CoordinateReferenceSystem nativeCRS = getSchema().getCoordinateReferenceSystem();
```
Comment (jodygarnett) [inline modules/library/data/src/main/java/org/geotools/data/store/ContentFeatureSource.java]: Nice clear code @tbarsballe, I am impressed you even updated the date on these files.

File: modules/library/data/src/test/java/org/geotools/data/memory/MemoryDataStoreBoundsTest.java [hunk 3]
```diff
@@ -53,8 +56,9 @@ protected void setUp() throws Exception {
         data = new MemoryDataStore();
         data.addFeatures(roadFeatures);
         
-        SimpleFeatureType riverTypeWithCrs = SimpleFeatureTypeBuilder.retype(riverType, DefaultGeographicCRS.WGS84);
+        SimpleFeatureType riverTypeWithCrs = SimpleFeatureTypeBuilder.retype(riverType, CRS.decode("EPSG:4326"));
```
Comment (jodygarnett) [inline modules/library/data/src/test/java/org/geotools/data/memory/MemoryDataStoreBoundsTest.java]: DefaultGeographicCRS.WGS84 and EPSG:326 represent the same idea mathematically. We *may* get a slight difference based on axis order settings.

I got no problem using EPSG:4326 here as it is consistent with testing reprojection code.

**Comment by jodygarnett** [comment]
Thanks @tbarsballe, quite impressed with your clarity on this one (great when reviewing a couple days before release).

File: modules/library/data/src/main/java/org/geotools/data/store/ContentFeatureSource.java (hunk 1 above)
Comment (tbarsballe) [inline modules/library/data/src/main/java/org/geotools/data/store/ContentFeatureSource.java]: Yeah, I just used the same structure and wording as in the reprojection section of getFeatures() for consistency.

File: modules/library/data/src/test/java/org/geotools/data/memory/MemoryDataStoreBoundsTest.java (hunk 3 above)
Comment (tbarsballe) [inline modules/library/data/src/test/java/org/geotools/data/memory/MemoryDataStoreBoundsTest.java]: Some of the comparaison tests were failing with the previous configuration (Specifically assertEquals against the features and the CRSs). While they may be the same mathematically, they have slightly different WKT representations. Consequently, I decided to stick with EPSG:4326 for consistency.

**Pull request merged**